tmol/pack/rotamer/build_rotamers.py

```diff
@@ -0,0 +1,1043 @@
+from tmol.types.functional import validate_args
+
+from tmol.utility.tensor.common_operations import exclusive_cumsum1d, stretch
+from tmol.database.chemical import ChemicalDatabase
+from tmol.kinematics.datatypes import KinForest
+from tmol.kinematics.compiled.compiled_ops import forward_only_op
+from tmol.chemical.restypes import RefinedResidueType
+from tmol.pose.packed_block_types import PackedBlockTypes
+from tmol.pose.pose_stack import PoseStack
+from tmol.pack.packer_task import PackerTask
+from tmol.pack.rotamer.chi_sampler import ChiSampler
+
+from tmol.pack.rotamer.single_residue_kinforest import (
+    construct_single_residue_kinforest,
+    coalesce_single_residue_kinforests,
+)
+from tmol.pack.rotamer.mainchain_fingerprint import (
+    annotate_residue_type_with_sampler_fingerprints,
+    find_unique_fingerprints,
+)
+
+
+def exc_cumsum_from_inc_cumsum(cumsum):
+    return numpy.concatenate((numpy.zeros(1, dtype=numpy.int64), cumsum[:-1]))
+
+
+@attr.s(auto_attribs=True, slots=True, frozen=True)
+class RotamerSet(ValidateAttrs):
+    n_rots_for_pose: Tensor[torch.int64][:]
+    rot_offset_for_pose: Tensor[torch.int64][:]
+    n_rots_for_block: Tensor[torch.int64][:, :]
+    rot_offset_for_block: Tensor[torch.int64][:, :]
+    pose_for_rot: Tensor[torch.int64][:]
+    block_type_ind_for_rot: Tensor[torch.int64][:]
+    block_ind_for_rot: Tensor[torch.int32][:]
+    coords: Tensor[torch.float32][:, :, :]
+
+
+# from tmol.system.restype import RefinedResidueType
+
+
+# step 1: let the dunbrack library annotate the residue types
+# step 2: let the dunbrack library annotate the condensed residue types
+# step 3: flatten poses
+# step 4: use the chi sampler to get the chi samples for all poses
+# step 5: count the number of rotamers per pose
+# step 5a: including rotamers that the dunbrack sampler does not provide (e.g. gly)
+# step 6: allocate a n_poses x max_n_rotamers x max_n_atoms x 3 tensor
+# step 7: create (n_poses * max_n_rotamers * max_n_atoms) x 3 view of coord tensor
+# step 8: create parent indexing based on start-position offset + residue-type tree data
+# step 9: build kinforest
+# step 10: take starting coordinates from residue roots
+# step 10a: take internal dofs from mainchain atoms
+# step 10b: take internal dofs for other atoms from rt icoors
+# step 11: refold
+
+
+# residue type annotations:
+#  - icoors
+#  - parent index (i.e. tree)
+#  - list of atoms whose icoors are controlled by dunbrack
+#  - bonds that are part of named torsions
+#  - mainchain atoms
+#  - rottable set for this RT
+#
+# condensed residue type annotations:
+#  - (all data in residue-type annotations)
+#
+#
+# NOTE: sidechain samplers should be listed in the residue-level packer task
+# NOTE: all samplers should be given the same starting data and then a mask
+#       to indicate which residues they should operate on
+
+
+@validate_args
+def rebuild_poses_if_necessary(
+    poses: PoseStack, task: PackerTask
```

Comment:
is it necessary to rebuild the pose for this functionality?  Can PBT either be held separate from pose or updated in a way that does not change derived data w/i posestack?

Reply:
This is the consequence of two design decisions:

The first is a rule for the `PackedBlockTypes` class is that new ResidueTypes / BlockTypes cannot be added to it after its creation. This allows annotations to be made to the PackedBlockTypes by name. E.g. Lennard Jones, by way of its base-class, the `AtomTypeDependentTerm`, makes the `atom_type` annotation on the PackedBlockType class, and then, when setting up for scoring, only checks if the annotation exists, but does not do any consistency checks.

The second is that `PackedBlockTypes` should be a data member of the `PoseStack` (as opposed to, say, global data), and the residue-type indices of the `PoseStack` depend on what block types exist in the `PoseStack`'s PackedBlockTypes object. Thus if there are only Leu block types in the PBT object before packing, and the packer wants to introduce other block types, then it needs to rebuild the pose with this alternate PBT. This is easy, however: besides the `block_type_ind`/`block_type_ind64 `tensors, the other tensors need only be reallocated and data copied into them.

Now, there's no rule that says that we can't initialize a `PoseStack` with the One True PBT that contained all the BTs needed for a particular trajectory. In that case, the rebuilding would be unnecessary. However, if we want to allow users to expand the set of BTs they allow over the course of a trajectory, then we would need to ensure the packer has the machinery necessary to rebuild the `PoseStack`.

Reply:
OK, that makes sense.